### PR TITLE
fix(client): use DELETE /users/:userId for account deletion

### PR DESCRIPTION
**Pull request by Sembauke**
Checklist:

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/how-to-open-a-pull-request/).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitHub Codespaces.

Ref #50301

### CODE REVIEW

**Comment by Qodo-Free-For-OSS** [comment]
Hi, deleteAccountSaga always shows the success flash and clears user state even if DELETE /users/:userId fails, because the ajax request helper does not throw on non-2xx and the saga never checks response.ok/status.

**Severity:** action required | **Category:** correctness

**How to fix:** Check response.ok before success

**Agent prompt to fix** - you can give this to your LLM of choice:

> ### Issue description
> `deleteAccountSaga` treats any HTTP response as success and proceeds to show `FlashMessages.AccountDeleted` and clear local user state.
> 
> ### Issue Context
> The ajax `request` helper does not throw on non-2xx responses, so callers must check status/ok or the helper must enforce it.
> 
> ### Fix Focus Areas
> - client/src/redux/settings/danger-zone-saga.js[12-34]
> - client/src/utils/ajax.ts[54-61]
> - client/src/utils/ajax.ts[84-101]
> 
> ### What to change
> - Option A (localized): capture `{ response }` from `yield call(postDeleteAccount, userId)` and only proceed when `response.ok` (or `response.status === 204`). Otherwise throw or dispatch an error/flash.
> - Option B (systemic): update `request/combineDataWithResponse` to throw (or wrap) for non-2xx responses so sagas go to `catch` automatically.

---
*Found by [Qodo](https://github.com/marketplace/qodo-merge-pro-for-open-source). Free code review for open-source maintainers.*

File: client/src/utils/ajax.ts [hunk 1]
```diff
@@ -52,6 +52,10 @@ async function get<T>(
 }
 
 async function combineDataWithResponse<T>(response: Response) {
+  if (response.status === 204 || response.status === 205) {
```
Comment (ojeytonwilliams) [inline client/src/utils/ajax.ts]:
```suggestion
  if (!response.body) {
```
Status codes are more like guidelines. If it's a 204 it shouldn't return anything, but who knows.

This is masking a deeper problem, namely that `combineDataWithResponse` was a silly idea (don't check git blame) and applying it to the delete endpoints (which either don't, or shouldn't, return anything) was sillier.

Anyways, for now, I think this is fine.

**Review by ojeytonwilliams — DISMISSED** [review]

**Comment by ojeytonwilliams** [comment]
Hey Sem, it looks like the e2e tests are surfacing a real error. Could you take a look?

File: client/src/redux/settings/danger-zone-saga.js [hunk 2]
```diff
@@ -1,29 +1,33 @@
...
-function* deleteAccountSaga() {
+export function* deleteAccountSaga() {
   try {
-    yield call(postDeleteAccount);
+    const userId = yield select(userIdSelector);
+    if (!userId) {
+      throw new Error('Unable to delete account: no user id found');
+    }
+    yield call(postDeleteAccount, userId);
+    yield call(getSignout);
```
Comment (ShaunSHamilton) [inline client/src/redux/settings/danger-zone-saga.js]: This should not be necessary, as both `DELETE /user/...` and `/signout` call `req.clearOurCookies`

File: client/src/redux/settings/danger-zone-saga.js [hunk 3]
```diff
@@ -1,29 +1,33 @@
...
         message: FlashMessages.AccountDeleted
       })
     );
-    // Navigate before signing out, since /settings will attempt to sign users
-    // back in if resetUserData fires while still on /settings.
-    void navigate('/learn');
-    // Wait for Gatsby to complete the route transition before clearing user
-    // data, ensuring /settings is unmounted and won't re-authenticate.
-    yield take(appTypes.routeUpdated);
-    yield put(resetUserData());
```
Comment (ShaunSHamilton) [inline client/src/redux/settings/danger-zone-saga.js:26]: I would have thought this would still be needed?

**Review by ShaunSHamilton — CHANGES_REQUESTED** [review]

File: client/src/redux/settings/danger-zone-saga.js (hunk 3 above)
Comment (Sembauke) [reply client/src/redux/settings/danger-zone-saga.js:26]: I think this one's fine, actually. Old code used Gatsby's `navigate('/learn')`, a client-side route change, so the app and Redux store stayed alive and `resetUserData()` was doing real work clearing stale state.

This version uses `hardGoTo('/learn')` instead, which is just `window.location.href = '/learn'` under the hood (see hard-go-to-epic.js). That's a full page reload, so everything gets torn down anyway, nothing left to reset.

**Review by ShaunSHamilton — APPROVED** [review]
The `combineDataWithResponse` changes took me for a rabbit-hole.

👍
